Task Testing

Andreyinthesky opened the pull request.

Commit:
Изменен ObjectComparison.cs: Перепсиан тест с исп. FA. Добавлен
комментарий по альтернативному решению

Commit:
Изменен NumberValidatorTests.cs: Переписаны тесты в соотв. с заданием.
Добавлено разделение тест кейсов на категории

EmptyBucket commented on cs/HomeExercises/ObjectComparison.cs:
```diff
@@ -15,17 +15,11 @@ public void CheckCurrentTsar()
-			Assert.AreEqual(expectedTsar.Parent.Parent, actualTsar.Parent.Parent);
-		}
+            // Перепишите код на использование Fluent Assertions.
+			actualTsar.ShouldBeEquivalentTo(expectedTsar, options => options.Excluding(p => p.Id)
```
Попробуй переписать это так, чтобы можно было безболезненно добавлять парентов и это не крашило тест

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -8,27 +9,87 @@ namespace HomeExercises
+		{
+			if (doesThrow)
+			{
+				Assert.That(() => new NumberValidator(precision, scale, onlyPositive),
```
Примени FluentAssertions

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -8,27 +9,87 @@ namespace HomeExercises
 	{
 		[Test]
-		public void Test()
+		[TestCaseSource(nameof(NumberValidatorTestCases_StringValue_ShouldBeMatchRegex), Category =
```
Поработай над именами тестов. См. "Тесты как спецификация". Посмотри практики именования тестов

![default](https://user-images.githubusercontent.com/8377311/47661294-92514600-dbba-11e8-9f38-c525840bbf00.png)

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -8,27 +9,87 @@ namespace HomeExercises
+		[TestCase(1, 0, true, false)]
+		public void TestThrows(int precision, int scale, bool onlyPositive, bool doesThrow)
+		{
+			if (doesThrow)
```
Разбивай на пару тестов и именовать проще будет и логический оператор уберешь

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -8,27 +9,87 @@ namespace HomeExercises
+		}
+
+		[Category("TestThrows")]
+		[TestCase(-1, 2, true, true)]
```
Не все кейсы покрыты. Иди по ветвлениям, делай тест на каждый предикат в конуструкторе

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -8,27 +9,87 @@ namespace HomeExercises
 		}
-	}
+
+		public static IEnumerable NumberValidatorTestCases_StringValue_ShouldBeMatchRegex
```
Данные можно унести в класс хранилку или скрыть region, чтобы не создавать шум в тесте, либо декларативно описать в атрибутах

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -8,27 +9,87 @@ namespace HomeExercises
+			}
+		}
+
+		public static IEnumerable NumberValidatorTestCases_IntAndFracPartCountSigns
```
Обрати внимание какие бывают разделители, так-же попробуй несколько разделителей в тексте, посмотри куда поставить разделитель, чтобы поломать

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -8,27 +9,87 @@ namespace HomeExercises
+			get { yield return new TestCaseData(17, 2, false, "0.000").Returns(false); }
+		}
+
+		public static IEnumerable NumberValidatorTestCases_NumberSign
```
Так-же можно попробовать несколько знаков, попробовать их на другое место поставить. В общем, последовательно пройдись по каждой группе в регулярке и попробуй ее поломать

Andreyinthesky commented:
comment

Commit:
Update ObjectComparison.cs: fix test

Commit:
Update NumberValidatorTests.cs
added new test cases, test are divided, FA applied

Commit:
Update NumberValidatorTests.cs
added new test cases, test are divided, FA applied

Commit:
fix NumberValidatorTests.cs

EmptyBucket commented on cs/HomeExercises/ObjectComparison.cs:
```diff
@@ -16,16 +17,10 @@ public void CheckCurrentTsar()
-			Assert.AreEqual(expectedTsar.Parent.Parent, actualTsar.Parent.Parent);
-		}
+			actualTsar.ShouldBeEquivalentTo(expectedTsar, options => options
+				.Excluding(pi => pi.SelectedMemberPath.EndsWith("Id"))
```
можно сделать .SelectedMemberInfo.Name.Equals. Хардкодить имя свойства - плохо, можно поломать ренеймом, да и навигация не сработает. Посмотри в сторону nameof

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -8,27 +9,76 @@ namespace HomeExercises
+		[TestCase(3, 2, false, "-", ExpectedResult = false, TestName = "when value is minus")]
+		[TestCase(3, 2, false, ",", ExpectedResult = false, TestName = "when value is comma")]
+		[TestCase(3, 2, false, ".", ExpectedResult = false, TestName = "when value is decimal point")]
+		public bool IsValidNumber_NotValidateNumber(int precision, int scale, bool onlyPositive, string value)
```
В тестах с конструктором в имени правильно описал _ОжидаемыйРезультат. Тут описываешь поведение. Так-же ExpectedResult не имеет смысла, тк он одинаковый для всех кейсов, только осложнит поддержку. Нет проверок с целыми числами

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -8,27 +9,76 @@ namespace HomeExercises
+		[TestCase(17, 2, true, "0", ExpectedResult = true, TestName = "when value without scale")]
+		[TestCase(17, 2, true, "0.0", ExpectedResult = true, TestName = "when value length less than precision")]
+		[TestCase(3, 2, false, "+5.5", ExpectedResult = true, TestName = "when value have plus")]
+		[TestCase(3, 2, false, "-5.5", ExpectedResult = true, TestName = "when value have minus")]
```
Не все кейсы покрыты. Например, нету теста на знак и целая часть. Последовательно иди по группам в регулярке и пиши кейсы. OnlyPositive только на 0 протестирован. Не пощупал граничные кейсы

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -8,27 +9,76 @@ namespace HomeExercises
+
+		[Test]
+		[TestCase(1, 0, true, TestName = "when scale is zero")]
+		[TestCase(3, 2, true, TestName = "when scale is less than precision")]
```
Нет теста на конструктор по умолчанию

Commit:
Update ObjectComparison.cs
Replaced EndWith() on Equals(), added using nameof

Commit:
Update NumberValidatorTests.cs
Fix constructor class, added more test cases, removed unnessasary cases,
added exceptionMessage assertion

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -3,81 +3,125 @@
-		[TestCase(3, 2, false, "-5.5", ExpectedResult = true, TestName = "when value have minus")]
-		public bool IsValidNumber_ValidateNumber(int precision, int scale, bool onlyPositive, string value)
+		[TestCase(17, 2, "0.000", TestName = "frac part length greater than scale")]
+		[TestCase(17, 2, "-0.000", TestName = "frac part length greater than scale with plus")]
```
Местами перепутал имя

EmptyBucket commented on cs/HomeExercises/NumberValidatorTests.cs:
```diff
@@ -3,81 +3,125 @@
+
+		[TestCase(2, 0, "-1", TestName = "value is negative integer")]
+		[TestCase(3, 1, "-1.1", TestName = "value is negative decimal")]
+		public void IsValidNumber_WhenValueIsNegativeAsOnlyPositive_ReturnFalse(int precision, int scale, string value)
```
OnlyPositive только на результат с false протестирован. Вполне может быть, что код написали так, что добавление этого флага всегда возвращает false, поэтому нужно и на результат с true проверить

Commit:
Update NumberValidatorTests.cs
added new test cases, fixed old test cases

Commit:
Update NumberValidatorTests.cs: fix text format
